### ISSUE TITLE
Layout.name duplicates Cell.name; usage unclear

### ISSUE THREAD

**Issue by growly** [Vlsir/schema-proto#5]
We define a cell's name both in the `Cell` message and in the `Layout` message. When `Layout`s are coupled with `Cell`s (usually?) this is duplicate information. It's not clear which one should be set, or if users should expect the cell name to be taken from the `Cell` instead of the `Layout` when the latter is unset.

As an example, `gds21` relies on the `Layout.name` to write `GdsStruct`s, ignoring the `Cell` name field.

**Comment by dan-fritchman** [comment]
Yeah we should clarify and perhaps reorganize this a bit. 
There's a similar relationship between `Cell.name` and `circuit.Module.name`.  

We certainly have cases where we use `Module` "on its own", without a `Cell`. Notably everything touching netlisting, simulation, HDL. (Really everything that actually uses `Module`.) `Layout` has similarly "operated independently" of `Cell` in many versions of Layout21, although not the current one. Maybe we don't want it to any more. 

More generally, I think of each of the fields of `Cell` as a "view": 

* Module (circuit implementation) 
* Interface (circuit abstract/ IO)
* Layout (physical implementation)
* (Physical / Layout) Abstract 

If any of these things is ever going to operate on their own, it would seem they need to either: 

* (a) Have the cell-name as a field, or 
* (b) Have some sort of "pointer" back to their parent `Cell`, which has the cell-name. 

We could make "parent pointers" like (b) would work in proto-land, but I think only with a bit of an overhaul. I think the better option is (a).  

Now, if there are such sometimes-redundant names, it certainly appears the rules for when those names differ are not terribly well-defined. They should be. The options would seem to be: 

* (1) Have a preference, either for the "higher" or "lower" level names, or 
* (2) Generate errors

Not sure I have a strong preference between the two. For your own recent work, it seems that (2) would've helped avoid some problems. 

---

Semi-separately - our organizing `Cell` into `layout.raw` doesn't seem quite right either. That multi-view concept seems to deserve its own space, given it's got equal parts "logical" and "physical" (layout) info: 

```
// # Cell 
// A multi-view representation of a piece of hardware.
message Cell {
  // Cell Name
  string name = 1;

  // IO Interface
  vlsir.circuit.Interface interface = 10;
  // Circuit Module 
  vlsir.circuit.Module module = 11;
  // Physical Abstract
  Abstract abstract = 12;
  // Physical Layout Implementation
  Layout layout = 13;
}
```

And we organize `layout.raw.Library` as a collection of these things

```
// # Library 
// A collection of cells and asssociated meta-data. 
message Library {
  // Library Name / Domain 
  string domain = 1;
  // Distance Units
  Units units = 2; 
  // Cell Definitions
  repeated Cell cells = 10;

  // Author Information 
  vlsir.utils.AuthorMetadata author = 20;
} 
```

**Comment by growly** [comment]
Yeah, agreed. I'd add a third option for dealing with Cell "Views": we can require that every "View" come wrapped in a `Cell` even if there are no other Views with it. This would preserve the tree-like structure of the data, avoid any back referencing and add only trivial overhead to the wire data. The major downside would be that the interfaces would be slightly less obvious. Luckily we only have to worry about that when ingesting/excreting (is that the opposite word?) protos.

If not that then I'd prefer your option (a). We just have to implement a defaulting behaviour (if Layout/Module name not set use Cell name) and stipulate that this is expected.

I also agree that `Cell` should move and each "View" message should sit at the same level of the namespace hierarchy.
